src/libsyntax/parse/parser.rs

```diff
@@ -1207,48 +1206,33 @@ impl<'a> Parser<'a> {
         Ok(())
     }
 
-    /// Parses `extern` followed by an optional ABI string, or nothing.
-    fn parse_extern_abi(&mut self) -> PResult<'a, Abi> {
-        if self.eat_keyword(kw::Extern) {
-            Ok(self.parse_opt_abi()?.unwrap_or(Abi::C))
+    /// Parses `extern string_literal?`.
+    /// If `extern` is not found, the Rust ABI is used.
+    /// If `extern` is found and a `string_literal` does not follow, the C ABI is used.
+    fn parse_extern_abi(&mut self) -> PResult<'a, Spanned<Abi>> {
+        Ok(if self.eat_keyword(kw::Extern) {
+            self.parse_opt_abi(self.prev_span)
         } else {
-            Ok(Abi::Rust)
-        }
+            dummy_spanned(Abi::RUST)
+        })
     }
 
-    /// Parses a string as an ABI spec on an extern type or module. Consumes
-    /// the `extern` keyword, if one is found.
-    fn parse_opt_abi(&mut self) -> PResult<'a, Option<Abi>> {
-        match self.token.kind {
-            token::Literal(token::Lit { kind: token::Str, symbol, suffix }) |
-            token::Literal(token::Lit { kind: token::StrRaw(..), symbol, suffix }) => {
-                self.expect_no_suffix(self.token.span, "an ABI spec", suffix);
-                self.bump();
-                match abi::lookup(&symbol.as_str()) {
-                    Some(abi) => Ok(Some(abi)),
-                    None => {
-                        self.error_on_invalid_abi(symbol);
-                        Ok(None)
-                    }
+    /// Parses a string literal as an ABI spec.
+    /// If one is not found, the "C" ABI is used with the provided span.
+    fn parse_opt_abi(&mut self, span: Span) -> Spanned<Abi> {
+        let abi: Option<_> = try {
+            match self.token.literal()? {
+                token::Lit { kind: token::Str, symbol, suffix } |
+                token::Lit { kind: token::StrRaw(..), symbol, suffix } => {
+                    let symbol = *symbol;
+                    self.expect_no_suffix(self.token.span, "an ABI spec", suffix.clone());
+                    self.bump();
+                    respan(self.prev_span, Abi { symbol })
                 }
+                _ => None?,
```

Comment:
Not a bug; the lack of `self.bump();` in this branch means that it will error on the next token. However, we could use better recovery, so I'll make some changes.